docs/HEADROOM_FOR_MPPOD.md

```diff
@@ -0,0 +1,79 @@
+# [Experimental] Reserving headroom for Mountpoint Pods
+
+_This is an experimental feature and the API or behavior may change in subsequent MINOR [releases](https://github.com/awslabs/mountpoint-s3-csi-driver?tab=readme-ov-file#releases)_.
+
+This feature causes overprovisioning in [node autoscalers](https://kubernetes.io/docs/concepts/cluster-administration/node-autoscaling/) like [Karpenter](https://karpenter.sh/) with the goal of reserving some headroom for _dynamically_ spawned Mountpoint Pods _after_ node creation.
+
+## Why is it needed?
+
+The CSI Driver v2 spawns Mountpoint Pods into the same nodes as the workloads to provide volumes. The CSI Driver spawns these Mountpoint Pods _after_ the workloads are assigned to nodes. This approach is taken because we might be able to [share an existing Mountpoint Pod](https://github.com/awslabs/mountpoint-s3-csi-driver/blob/v2/docs/MOUNTPOINT_POD_SHARING.md) on a specific node, eliminating the need to spawn a new one. Therefore, the CSI Driver needs to wait until a workload is scheduled to a specific node before deciding whether to share an existing Mountpoint Pod or spawn a new one.
+
+This dynamic spawning of Mountpoint Pods on specific nodes creates a problem with node autoscalers like Karpenter. Since autoscalers are not aware of Mountpoint Pod requirements when making scaling decisions, they may make suboptimal choices. As a result, there may not be enough space for a Mountpoint Pod on the specific node where the workload is deployed.
+
+This feature aims to mitigate this issue by reserving headroom for upcoming Mountpoint Pods, ensuring autoscalers create instances large enough to host Mountpoint Pods alongside workloads.
+
+## How does it work?
+
+When the CSI Driver detects a Workload Pod using the scheduling gate to enable this feature, it performs the following steps:
+
+  1. Labels the Workload Pod to use inter-pod affinity rules in the Headroom Pods
+  2. Creates Headroom Pods using a pause container with inter-pod affinity rule to the Workload Pod - since node autoscalers like [Karpenter supports inter-pod affinity rules](https://karpenter.sh/docs/concepts/scheduling/#pod-affinityanti-affinity), this should help them to choose a right instance type
+  3. Ungates the scheduling gate from the Workload Pod to let it scheduled - alongside the Headroom Pods if possible
+  4. Schedules Mountpoint Pod if necessary (i.e., the CSI Driver cannot share an existing Mountpoint Pod) into the same node as the Workload and Headroom Pods using a preempting priority class
+  5. Mountpoint Pod most likely preempts the Headroom Pods if there is no space in the node - as the Headroom Pods uses a negative priority -, or just gets scheduled if there is enough space for all pods
+  6. Deletes the Headroom Pods as soon as the Workload Pod is running or terminated - as Mountpoint Pods are already scheduled or no longer needed
```

Comment:
> Deletes the Headroom Pods as soon as the Workload Pod is running or terminated

Sometimes Reconciler might miss that Workload Pod is terminated due to eventual consistency and we will log `Pod not found - ignoring` from line 78 in reconciler.go.

In this case, Headroom Pods won't be deleted, right?

I.e. we would see ``

Reply:
Yeah that's true. We also have the same issue with Mountpoint Pods I think, we need to have a periodic cleanup for missed events.